### PR TITLE
issue #418 Move gen_c_code() methods to C/OpenCL back-ends

### PR DESCRIPTION
**Pull request by sergisiso**
This PR deletes the methods that generated C code or OpenCL code from the PSyIR nodes and moves the logic into two new psyir backends following the VisitorPattern. Note that the remaining gen_ocl* methods in the nodes are the ones that generate the OpenCL psy layer in Fortran.

Additionally some gocean opencl generation has been generalized, so now only a few abstract methods of the SymbolTable are required to be implemented by an API in order to produce kernel opencl.

### CODE REVIEW

**Comment by sergisiso** [comment]
@rupertford @arporter If all test pass this is ready for review.

File: src/psyclone/psyGen.py [hunk 1]
```diff
@@ -5819,7 +5715,14 @@ def lhs(self):
         :returns: the child node representing the Left-Hand Side of the \
             assignment.
         :rtype: :py:class:`psyclone.psyGen.Node`
+
+        :raises InternalError: Node has lest children than expected
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py]: Done

File: src/psyclone/psyGen.py [hunk 2]
```diff
@@ -5819,7 +5715,14 @@ def lhs(self):
...
         '''
+        if not self._children:
+            raise InternalError(
+                "Assignment '{}' malformed or incomplete. It "
+                "needs at least 1 children to have a lhs.".format(str(self)))
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py]: done

File: src/psyclone/psyir/backend/c.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+class CWriter(PSyIRVisitor):
+    '''Implements a PSyIR-to-C back-end for the PSyIR AST.
+
+    '''
+
+    def gen_declaration(self, symbol):
+        '''
+        Generates string representing the C declaration of the symbol. In C
+        declarations can be found be inside the argument list or with the
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: ups, fixed

File: src/psyclone/psyir/backend/c.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+        statments, so no indention or punctuation is generated by this method.
+
+        :param symbol: The symbol instance.
+        :type symbol: :py:class:`psyclone.psyGen.Symbol`
+
+        :returns: The C declaration of the given of the symbol.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: fixed

File: src/psyclone/psyir/backend/c.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        '''This method is called when an Array instance is found in the PSyIR
+        tree.
+
+        :param node: An Array PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Array`
+
+        :returns: The C code as a string.
+        :rtype: str
+
+        :raises VisitorError: If this node has children.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: fixed

File: src/psyclone/psyir/backend/c.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+    def ifblock_node(self, node):
+        '''This method is called when an IfBlock instance is found in the
+        PSyIR tree.
+
+        :param node: An IfBlock PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.IfBlock`
+
+        :returns: The C code as a string.
+        :rtype: str
+
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: done

File: src/psyclone/psyir/backend/c.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        '''
+        if len(node.children) < 2:
+            raise VisitorError(
+                "IfBlock malformed or incomplete. It should have at least "
+                "2 children, but found {0}.".format(len(node.children)))
+
+        condition = self._visit(node.children[0])
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: indeed :)

File: src/psyclone/psyir/backend/c.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+        '''This method is called when a UnaryOperation instance is found in
+        the PSyIR tree.
+
+        :param node: A UnaryOperation PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.UnaryOperation`
+
+        :returns: The C code as a string.
+        :rtype: str
+
+        :raises VisitorError: If this node has more than one children.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: done

File: src/psyclone/psyir/backend/c.py [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+    def binaryoperation_node(self, node):
+        '''This method is called when a BinaryOperation instance is found in
+        the PSyIR tree.
+
+        :param node: A BinaryOperation PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.BinaryOperation`
+
+        :returns: The C code as a string.
+        :rtype: str
+
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: done

File: src/psyclone/psyir/backend/opencl.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+'''
+
+from psyclone.psyir.backend.base import VisitorError
+from psyclone.psyir.backend.c import CWriter
+
+
+class OpenCLWriter(CWriter):
+    '''Implements a PSyIR-to-OpenCL back-end for the PSyIR AST. This writer
+    procudes OpenCL following the specs of the Version 1.2.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/opencl.py]: fixed

File: src/psyclone/psyir/backend/opencl.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    (https://www.khronos.org/registry/OpenCL/specs/opencl-1.2.pdf)
+
+    '''
+
+    def gen_id_variable(self, symbol, dimension_index):
+        '''
+        Generate the declaration and initialisation of OpenCL work-items
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/opencl.py]: rephrased docstring

File: src/psyclone/psyir/backend/opencl.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        global id variables. This are initialised by the OpenCL function:
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/opencl.py]: rephrased

File: src/psyclone/psyir/backend/opencl.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+        code = ""
+        code += self._nindent + "int " + symbol.name
+        code += " = get_global_id(" + str(dimension_index) + ");\n"
+        return code
+
+    def gen_declaration(self, symbol):
+        '''
+        Generate the declaration of an OpenCL variable. Note this can
+        be local variables or arguments, so no indention or punctuation
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/opencl.py]: done

File: src/psyclone/psyir/backend/opencl.py [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+        '''
+        prefix = ""
+        if symbol.shape:
+            prefix += "__global "
+        return prefix + super(OpenCLWriter, self).gen_declaration(symbol)
+
+    def gen_array_length_variables(self, symbol, symtab=None):
+        '''
+        Generate length variable declaration and initialisation for each array
+        dimension of the given symbol. By convetion they are named:
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/opencl.py]: fixed

File: src/psyclone/tests/psyir/backend/c_test.py [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author S. Siso, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+'''Performs pytest tests on the psyclond.psyir.backend.c module'''
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/c_test.py]: fixed

File: src/psyclone/tests/psyir/backend/c_test.py [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+    '''
+    cwriter = CWriter()
+
+    monkeypatch.setattr(cwriter, "gen_declaration",
+                        lambda x: "<declaration>")
+
+    # Local variables are declared as single statements
+    symbol = Symbol("dummy1", "integer")
+    result = cwriter.gen_local_variable(symbol)
+    assert result == "<declaration>;\n"
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/c_test.py]: I added a comment. It checks that it has the mocked `gen_declaration` in a statement form (ends with `;\n`) The declaration itself is tested more extensively on `gen_declaration` test.

File: src/psyclone/tests/psyir/backend/c_test.py [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+    result = cwriter(assignment)
+    assert result == "a = b;\n"
+
+    # Now add a child to the reference node
+    assignment.lhs.addchild(Node(parent=assignment))
+
+    # Generate C from the PSyIR schedule
+    with pytest.raises(VisitorError) as excinfo:
+        result = cwriter(assignment)
+    assert "PSyIR Reference node should not have any children." in str(excinfo)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/c_test.py]: Added `str(node)` in the error message

File: src/psyclone/tests/psyir/backend/c_test.py [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+
+
+def test_cw_codeblock():
+    '''Check the CWriter class codeblock method raises the expected
+    error.
+    '''
+
+    cblock = CodeBlock([])
+
+    # Generate Fortran from the PSyIR schedule
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/c_test.py]: ups, deleted

File: src/psyclone/tests/psyir/backend/c_test.py [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+    cwriter = CWriter()
+
+    with pytest.raises(VisitorError) as error:
+        _ = cwriter(cblock)
+    assert "CodeBlocks can not be translated to C." in str(error)
+
+
+def test_cw_unaryoperator():
+    '''Check the CWriter class unary_operation method correctly prints out
+    the C representation. Uses -1 as the example.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/c_test.py]: modified docstring

File: src/psyclone/tests/psyir/backend/c_test.py [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+    unary_operation._operator = Unsupported
+    with pytest.raises(NotImplementedError) as err:
+        _ = cwriter(unary_operation)
+    assert "The C backend does not support the '" in str(err)
+    assert "' operator." in str(err)
+
+
+def test_cw_binaryoperator():
+    '''Check the CWriter class binary_operation method correctly
+    prints out the Fortran representation of an intrinsic.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/c_test.py]: modified docstring

File: src/psyclone/tests/psyir/backend/opencl_test.py [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+
+
+def test_oclw_gen_array_length_variables():
+    '''Check the OpenCLWriter class gen_array_length_variables method produces
+    the expected declarations.
+
+    '''
+    oclwriter = OpenCLWriter()
+
+    # An scalar should not return any LEN variables
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/opencl_test.py]: fixed

File: src/psyclone/tests/psyir/backend/opencl_test.py [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+        "int dummy2LEN2 = get_global_size(1);\n" \
+        "int dummy2LEN3 = get_global_size(2);\n"
+
+    # Create a symbol table
+    symtab = SymbolTable()
+    symtab.add(symbol1)
+    symtab.add(symbol2)
+    symtab.add(symbol3)
+
+    # If there are no name clashed, generatre array length variables.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/opencl_test.py]: fixed

File: src/psyclone/psyir/backend/c.py [hunk 23]
```diff
@@ -0,0 +1,400 @@
...
+        :raises NotImplementedError: if there are some symbol types or nodes \
+            which are not implemented yet.
+        '''
+        code = ""
+        if symbol.datatype == "real":
+            code = code + "double "
+        elif symbol.datatype == "integer":
+            code = code + "int "
+        elif symbol.datatype == "character":
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: added tests in `test_cw_gen_declaration`

File: src/psyclone/psyir/backend/c.py [hunk 24]
```diff
@@ -0,0 +1,400 @@
...
+        :rtype: str
+
+        :raises VisitorError: If this node has children.
+
+        '''
+        code = node.name + "["
+
+        dimensions_remaining = len(node.children)
+        if dimensions_remaining < 1:
+            raise VisitorError("Array must have at least 1 dimension.")
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: Added `str(node)` in the error message and a test for it.

File: src/psyclone/psyGen.py [hunk 25]
```diff
@@ -5819,7 +5715,14 @@ def lhs(self):
         :returns: the child node representing the Left-Hand Side of the \
             assignment.
         :rtype: :py:class:`psyclone.psyGen.Node`
+
+        :raises InternalError: Node has lest children than expected
         '''
+        if not self._children:
+            raise InternalError(
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py]: added test

File: src/psyclone/psyGen.py [hunk 26]
```diff
@@ -5828,7 +5731,14 @@ def rhs(self):
         :returns: the child node representing the Right-Hand Side of the \
             assignment.
         :rtype: :py:class:`psyclone.psyGen.Node`
+
+        :raises InternalError: Node has lest children than expected
         '''
+        if len(self._children) < 2:
+            raise InternalError(
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py]: added test

**Comment by sergisiso** [comment]
@arporter This is ready for the next review. I addressed in-line comments and increased the coverage. The documentation already has a good back-ends explanation and the OpenCL functionality has not changed. I just added a new developer's guide section listing the available back-ends and mentioning that the C back-end at the moment is just for partial expressions, not for full kernels.

File: src/psyclone/tests/psyGen_test.py [hunk 27]
```diff
@@ -2803,9 +2803,24 @@ def test_assignment_semantic_navigation():
     '''Test that the Assignment navigation properties reference the expected
     children'''
     assignment = Assignment()
+
+    # lhs should fail if first child is not present
+    with pytest.raises(InternalError) as err:
+        _ = assignment.lhs
+    print(err)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyGen_test.py]: removed

File: src/psyclone/psyir/backend/opencl.py [hunk 28]
```diff
@@ -77,8 +77,8 @@ def gen_id_variable(self, symbol, dimension_index):
 
     def gen_declaration(self, symbol):
         '''
-        Generate the declaration of an OpenCL variable. Note this can
-        be local variables or arguments, so no indention or punctuation
+        Generate the declaration of an OpenCL variable. This can be either
+        a local variables or a routine argument, so no indention or punctuation
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/opencl.py]: fixed

File: src/psyclone/psyir/backend/c.py [hunk 29]
```diff
@@ -309,6 +316,10 @@ def binaryoperation_node(self, node):
         :returns: The C code as a string.
         :rtype: str
 
+        :raises VisitorError: If this node has fewer children than expected.
+        :raises NotImplementedError: If the operator in not supported by the \
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: fixed

File: src/psyclone/psyir/backend/c.py [hunk 30]
```diff
@@ -240,7 +245,9 @@ def unaryoperation_node(self, node):
         :returns: The C code as a string.
         :rtype: str
 
-        :raises VisitorError: If this node has more than one children.
+        :raises VisitorError: If this node has more than one child.
+        :raises NotImplementedError: If the operator in not supported by the \
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: fixed

File: doc/developer_guide/developers.rst [hunk 31]
```diff
@@ -778,6 +778,18 @@ hide users from the use of the visitor pattern. This is the interface
...
+===================
+
+Currently, there are two back-ends capable to generating Kernel
+code (a KernelSchedule with all its children), these are:
+
+- `FortranWriter()` in `psyclone.psyir.backend.fortran`
+- `OpenCLWriter()` in `psyclone.psyir.backend.opencl`
+
+Additionally, there is a `psyclone.psyir.backend.c` back-end, but at the
+moment it is just capable of processing partial PSyIR expressions.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/developers.rst]: done

File: doc/developer_guide/developers.rst [hunk 32]
```diff
@@ -778,6 +778,18 @@ hide users from the use of the visitor pattern. This is the interface
 to expose to users (which is why `_visit` is used for the visitor
 method, rather than `visit`).
 
+Available back-ends
+===================
+
+Currently, there are two back-ends capable to generating Kernel
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/developers.rst]: fixed

**Comment by sergisiso** [comment]
@arporter This is ready for next review.

**Comment by sergisiso** [comment]
I just merged with master, will be ready for review if it passes the tests.

File: src/psyclone/tests/psyir/backend/c_test.py [hunk 33]
```diff
@@ -0,0 +1,374 @@
+# -----------------------------------------------------------------------------
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/c_test.py]: This file also has a few pylint errors.

File: src/psyclone/tests/psyir/backend/opencl_test.py [hunk 34]
```diff
@@ -0,0 +1,182 @@
+# -----------------------------------------------------------------------------
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/opencl_test.py]: This file has a couple of pylint errors.

File: src/psyclone/gocean1p0.py [hunk 35]
```diff
@@ -1942,131 +1941,35 @@ def _check_gocean_conformity(self):
...
-
-        :param indent: Depth of indent for the output string.
-        :return: OpenCL argument list for the Symbol Table.
-        :rtype: str
-        '''
-        self._check_gocean_conformity()
-
-        arglist = []
-        for symbol in self.argument_list[2:]:
-            prefix = Node.indent(indent)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/gocean1p0.py]: removed from import

File: src/psyclone/tests/gocean1p0_opencl_test.py [hunk 36]
```diff
@@ -234,141 +234,92 @@ def test_opencl_kernel_code_generation():
     the expected OpenCL code.
     '''
 
+    from psyclone.psyir.backend.opencl import OpenCLWriter
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/gocean1p0_opencl_test.py]: fixed 2 pytlint errors

File: src/psyclone/tests/psyir/backend/c_test.py (hunk 33 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/c_test.py]: fixed all but protected accesses

File: src/psyclone/tests/psyir/backend/opencl_test.py (hunk 34 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/opencl_test.py]: fixed

File: src/psyclone/psyir/backend/c.py [hunk 37]
```diff
@@ -0,0 +1,411 @@
...
+        :rtype: str
+
+        '''
+        lhs = self._visit(node.lhs)
+        rhs = self._visit(node.rhs)
+
+        result = "{0}{1} = {2};\n".format(self._nindent, lhs, rhs)
+        return result
+
+    def reference_node(self, node):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: we found the same issue in the Fortran back-end and decided to keep the `self` in this case to make all visitor methods consistent.

File: src/psyclone/psyir/backend/c.py [hunk 38]
```diff
@@ -0,0 +1,411 @@
...
+
+from psyclone.psyir.backend.base import PSyIRVisitor, VisitorError
+
+
+class CWriter(PSyIRVisitor):
+    '''Implements a PSyIR-to-C back-end for the PSyIR AST.
+
+    '''
+
+    def gen_declaration(self, symbol):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: same

File: src/psyclone/psyir/backend/c.py [hunk 39]
```diff
@@ -0,0 +1,411 @@
...
+            for dim in reversed(range(1, dimensions_remaining)):
+                dimstring = node.name + "LEN" + str(dim)
+                code = code + " * " + dimstring
+            dimensions_remaining = dimensions_remaining - 1
+            code = code + " + "
+
+        code = code[:-3] + "]"  # Delete last ' + ' and close bracket
+        return code
+
+    def literal_node(self, node):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: same

File: src/psyclone/psyir/backend/c.py [hunk 40]
```diff
@@ -0,0 +1,411 @@
...
+        :param node: A Return PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Return`
+
+        :returns: The C code as a string.
+        :rtype: str
+
+        '''
+        return "{0}return;\n".format(self._nindent)
+
+    def codeblock_node(self, node):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py]: same

**Comment by sergisiso** [comment]
@arporter Fixed several pylint issues. An exception is the visitor method unused `self`. We could convert them to a static methods but then some of the visitor would be static and some would be regular methods, so it seems better to just keep the self for consistency (we did the same in the Fortran back-end).

File: src/psyclone/psyir/backend/c.py (hunk 37 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/backend/c.py]: OK.

**Comment by arporter** [comment]
All requested changes have been made.
All tests pass (inc. with compilation).
Just waiting on final Travis runs...

**Pull request merged**